### PR TITLE
Adding a spin/kokkos atom style

### PR DESCRIPTION
**Pull request by julient31**
**Summary**

This PR contains three main features:

- adding a spin/kokkos atom style 
- correcting a small bug in pair_spin_exchange_biquadratic.cpp
- adding the possibility of running fix nve/spin on non-triclinic boxes

**Related Issue(s)**

None

**Author(s)**

Julien Tranchida (Sandia)
Stan Moore (Sandia)
Svetoslav Nikolov (Sandia)

Contact: julien.tranchida1@gmail.com

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

All changes should be completely transparent to the associated LAMMPS input files.

**Implementation Notes**

A change was made in verlet_kokkos.cpp: adding the fm and fm_long spin vectors to the associated force_clear. 

All spin validation tests (in examples/SPIN/test_problems) were adapted and  successfully ran to check the new atom style. 

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [x] One or more example input decks are included

**Further Information, Files, and Links**

### CODE REVIEW

**Comment by julient31** [comment]
@stanmoore1 Stan, sorry it took me forever to find time to finish that atom spin/kk code. 
The temperature issue we observed was coming from a force_clear issue. To correct it, I had to add a force_clear option in verlet/kk (to clear fm and fm_long, the magnetic torque vectors). If you could have a quick look at this, that would be very helpful! Thanks!

**Comment by julient31** [comment]
Hey @akohlmey and @stanmoore1 
I just wanted to mention something real quick, as it may be relevant for future kokkos atom/styles. 
The issue I experienced with the absence of force_clear in verlet/kk for my spin atom style would also occur for other atom styles that have an extraflag = 1. 
I know this is not the case of a lot of atom styles so that it may not be very important for now, but as I was chatting about it with Aidan, he told me you guys may be interested to know.

File: src/KOKKOS/kokkos_type.h
Comment (stanmoore1) [inline src/KOKKOS/kokkos_type.h:776]: @julient31 can you make this a general type `tdual_float_1d_4`? Then in the Spin style you can typedef it to a new name.

File: src/KOKKOS/kokkos_type.h
Comment (stanmoore1) [inline src/KOKKOS/kokkos_type.h:797]: Please use the existing `tdual_f_array` instead of adding new types. You can typedef it to `tdual_fm_long_array` in the Spin class.

**Review by stanmoore1 — CHANGES_REQUESTED** [review]

**Comment by stanmoore1** [comment]
> Hey @akohlmey and @stanmoore1
> I just wanted to mention something real quick, as it may be relevant for future kokkos atom/styles.
> The issue I experienced with the absence of force_clear in verlet/kk for my spin atom style would also occur for other atom styles that have an extraflag = 1.
> I know this is not the case of a lot of atom styles so that it may not be very important for now, but as I was chatting about it with Aidan, he told me you guys may be interested to know.

@julient31 yes this is a good catch, I'll take a look at the other styles.

**Comment by julient31** [comment]
Hey @stanmoore1 
I started working on the corrections you asked me. I left the former code commented for now. I'll clean everything up as soon as I have your verification and approval. I modified the following:
- in kokkos_type.h, I create a standard tdual_float_1d_4 and used it to typedef of my spin vectors 
- in kokkos_type.h, I used the standard tdual_f_array to typedef my two magnetic "force" vectors (fm and fm_long) 
- in atom_kokkos.h, I made sure the definition of the spin vectors (k_sp, k_fm and k_fm_long) match the definitions above,
- in same in atom_vec_spin_kokkos.cpp

Let me know if this is what you were expecting, or if you would like additional changes!
Best,
Julien.

**Comment by julient31** [comment]
Hey Stan @stanmoore1 !
Hope you're doing well. Did you have time to check those changes? 
As I said, I'll clean-up everything in the code (all the commented lines) as soon as I have your approval. Let me know!
Best,
Julien.

**Comment by stanmoore1** [comment]
@julient31 Yes this looks good. After you remove the comments it is ready for merging. Thanks

**Comment by julient31** [comment]
@stanmoore1 and @akohlmey 
I think it should be good now ;-). Otherwise, let me know.
Thanks a lot for the reviews!
Julien.

**Review by stanmoore1 — APPROVED** [review]

**Pull request merged**
